### ISSUE TITLE
Interface: M2M Extended suggestion

### ISSUE THREAD

**Issue by henrikwirth** [directus/app#1598]
# Feature Request

## What problem does this feature solve?

When creating an M2M, it would be a wonderful feature to be able to use the junction for extra information. So for example. I have a M2M-Relation like: Character <-> Item. I would love to be able to add the amount of Items.

| Field | Interface |
|---|---|
| ID  | Integer  | Primary Key  |
| character | Numeric  |
| item | Numeric |
| amount | Numeric |

Giving more opportunities for data modeling. The question of course here is... How to design the UI, so it is easy to enter these values, when adding the relation.

## How do you think this should be implemented?

I think it should be its own interface like: M2M Extended

**The way to input other fields could look something like this (of course better designed):**

![image](https://user-images.githubusercontent.com/10786260/57296945-8b0d7680-70ce-11e9-87f3-e342d1983608.png)

**And when adding an Item it could look like this:**

![image](https://user-images.githubusercontent.com/10786260/57297576-366afb00-70d0-11e9-95e3-75961754b412.png)

I think it would be very powerful. At first maybe just simple field types are accepted. But I guess you can theoretically use most of the interfaces. The only problem there is then you would need a proper way to fill the inputs of the relation somehow. Maybe a normal edit view in a popup would do the trick.

## Would you be willing to work on this?
I don't know, if I could do it alone. But If someone would give me a hand, I would be willing to try. 

I would love to hear what you think about this idea and also how difficult it is to create that interface based on the M2M-Interface?

**Comment by benhaynes** [comment]
Hey @henrikwirth — technically this is a duplicate (https://github.com/directus/app/issues/761) but your ticket has much more thought/details so I'll close the other one. Thank you for (re)submitting this, I've wanted this one for a while!

I have two thoughts:

### Junction Field Setup
I think it would be cleanest to re-use what we already have. By that I mean we could have admins go into the actual junction table from _Settings > Collections+Fields_ instead of trying to add the fields through the relationship modal. This already works and includes all of the extra info (datatype, default, required, etc)

### Editing Junction Data
Again, I think we want to re-use as much as possible (less code to build/maintain). My thinking here is that when you open up a related item to create/edit it, that you have a _Section_ for the junction fields, eg:

```
Item Name
Item Description
Item Image
-----------------
Amount
```

Again, then we can use _all_ interfaces without worrying about squishing them into a smaller interface. It also means that if we create 3-4 different M2M interfaces that all will work the same way, showing the junction fields in the related edit view.

So basically this feature is as "simple" as adding a conditional to the Item Create/Edit view that shows/saves the junction table fields (if they exist). Of course I'm probably oversimplifying how easy this is... right @rijkvanzanten? ;)

**Comment by henrikwirth** [comment]
Didn't see the other issue. Great, that this is already in the thought process. Your points surely make sense and why creating something new, when we can just reuse what we have. Makes also much more sense the way you suggested.

The only thing I don't understand yet, is how you actually add the values, when you create the relation. Showing them in the table is one thing, but where do you add them? If you could clarify that a bit more, maybe even visually, I would love to try developing it.

**Comment by benhaynes** [comment]
So in Directus there are essentially two main pages, a listing of items to browse a collection (`item browse`) and an item detail page used for viewing, creating, and editing individual items (`item detail`). This Item Detail page is a form comprised of fields, some of which are relational (eg: m2m).

So, if all that is clear, the flow would be:

1. Go to Item Detail `a`
2. Scroll to the M2M field and click "Create New"
    * You could also "Choose Existing Item" from that interface, and then once added you could open it to edit all the fields.
3. This opens a modal overlay with a new Item Detail form for creating the new relational item `b`
4. Within this modal we would have both the related item's fields as well as the junction table's fields (organized and labeled intuitively).

Alternatively, a new interface could be created where all of this is edited "inline" instead of in a modal. That would be similar in design/concept to this: https://github.com/directus/app/issues/1205#issuecomment-476266730

Does that make more sense?

**Comment by rijkvanzanten** [comment]
To achieve better clarity/visibility, we are now tracking feature requests within the [Feature Request](https://github.com/directus/app/projects/45) project board.

**This issue being closed does not mean it's not being considered.**

**Comment by rbarszcz** [comment]
so.. this was implemented, dropped, postponed.. any news?

**Comment by benhaynes** [comment]
In the upcoming v9, you'll be able to add fields to the junction table and they will show up as editable at the top of the related item form. Should have an Alpha for Directus 9 out this week.
